### ISSUE TITLE
Load resultset data given a resultset identifier

### ISSUE THREAD

**Issue by Alexander-Morin** [PavlidisLab/Gemma#127]
Given a GEO identifier and resultset identifier, the requested functionality would return the affiliated resultset table. In reality, the resultset identifier alone could be enough, but because it is not stable it seems prudent that the user should have to specify the GEO identifier as well. Then, if an "incorrect" call is made because the user supplied a once valid resultset identifier, the returned error message could list all of the resultsets identifiers and their respective factors for the given GEO identifier.

This call would need to be aware of cases where a GEO accession has been split, e.g:
https://gemma.msl.ubc.ca/expressionExperiment/showExpressionExperiment.html?shortName=GSE75650.1
https://gemma.msl.ubc.ca/expressionExperiment/showExpressionExperiment.html?shortName=GSE75650.2
(The second doesn't actually have any live resultsets at the moment, but you get the idea)  

The overall goal is that this call would follow the functionality requested in https://github.com/PavlidisLab/Gemma/issues/126, such that resultset identifiers can be retrieved and automatically fed into a loading call.

@arteymix

**Comment by arteymix** [comment]
I'm not exactly sure that Gemma keeps track of experiment splitting, but it at least keep tracks of external identifiers. What we could do in this case is to add the ability of querying result sets based on an external ID (source + accession). It would naturally include split experiments.

I agree here that if result set IDs are not constant (as more analysis can be performed or simply redone), it is not an appropriate way to query data. However, for your purposes, I would recommend that once you have what you need, you stick to those IDs for reproducibility.

**Comment by ppavlidis** [comment]
There are multiple ways you can figure out splits

1) ExpressionExperiment.otherParts is a collection, you can use that when you have one part to check for others.
2) An ExpressionExperimentSet is created that contains the parts, although experiments can be part of >1 EESet. I put this in as a kind of backup, it might not be useful.
3) And as @arteymix  says the GSE number is still associated with each part, so if you have the GSE you can find all associated parts (splitting by species works the same way)

See SplitExperimentServiceImpl. 

That reminds me that the current implementation doesn't delete the original experiment, partly just to guard against this somewhat experimental feature screwing up.

**Comment by Alexander-Morin** [comment]
I just want to clarify because it seems to have created a source of confusion: the focus of this issue is about loading resultsets based upon a resultset ID. The comment about splitting experiments is a side point that I raised because I imagined it could lead to complications, but in reality split experiments are rare. I don't want that to distract from the main requested functionality.

**Comment by arteymix** [comment]
I don't understand how using a result set identifier to retrieve the associated result sets is relevant to your query.

If you have a GEO identifier, then what you are interested into is all the result sets associated to the EE that share that external identifier. It will include the resultset you have in hand.

Affiliation can take a multiple meaning. It could be a shared ExpressionExperiment (via its ID or short name), or all those associated to an ExpressionExperimentSet (i.e. all the parts of a split experiment).

Here's the breakdown of the search criteria that I can implement:

 - by ExpressionExperiment identifier (generally results in a single ResultSet, but could be many if there has been separate DE analyses performed)
 - by EE short name
 - by an external identifier (would retrieve parts of a split experiment, or an experiment broken down by taxon)
 - by ExpressionExperimentSet identifier (would retrieve ResultSet associated to a bunch of expression experiments)

In any of the three cases, you need to explicitly decide which result sets you decide to keep. For that purpose, I will expand to include the associated EE metadata.

They would all constitute an extension to the "Dataset differential analysis"  endpoint.

**Comment by arteymix** [comment]
Retrieving result sets associated to a split is just a side effect of querying by a GEO identifier since they all share it. I don't really care about this aspect in the implementation.

**Comment by Alexander-Morin** [comment]
I'm worried that we've gone off the rails here:

My clarification message from earlier was in reference to an external thread that referenced this issue, not your message from 10 days ago Guillaume. 

> "I don't understand how using a result set identifier to retrieve the associated result sets is relevant to your query."

To be fair, this is the exact purpose of this issue (hence the title).

All of the following points you make regarding tracking experiment information is useful, but seems better suited to https://github.com/PavlidisLab/Gemma/issues/126

Sorry for causing/continuing any confusion.

**Comment by arteymix** [comment]
I will need a clear definition of what affiliated result sets are. Is it just a matter of deriving from the same EE? or sharing an external identifier like a GEO series?

Knowing this, it just becomes a matter of applying one of the filtering criteria I outlined.

For example, if your notion of affiliation is sharing the same EE (which makes most sense to me), I would first retrieve the EE associated to your result set then list all the sets that are associated to that EE.

The API would simply expose a new request parameter allowing you to filter by result set in addition to the other criteria I will implement.

**Comment by Alexander-Morin** [comment]
> I will need a clear definition of what affiliated result sets are. Is it just a matter of deriving from the same EE? or sharing an external identifier like a GEO series?
> 
> Knowing this, it just becomes a matter of applying one of the filtering criteria I outlined.
> 
> For example, if your notion of affiliation is sharing the same EE (which makes most sense to me), I would first retrieve the EE associated to your result set then list all the sets that are associated to that EE.
> 
> The API would simply expose a new request parameter allowing you to filter by result set in addition to the other criteria I will implement.



Retrieving all of Gemma' resultset identifiers for a given GSE identifier is the first priority, and is what is requested in: https://github.com/PavlidisLab/Gemma/issues/126

The request of this issue is to gain access to a singular resultset table upon submission of a GSE identifier and a singular resultset identifier, the latter which is assumed to be retrieved downstream of the functionality requested in https://github.com/PavlidisLab/Gemma/issues/126

**Comment by arteymix** [comment]
Ohhh! That makes much more sense.

I was interpreting this similarly to how SRA runs work, where you can use a SRR accession to retrieve an experiment which will list all the associated runs. But what you mean by result set table is the actual table of fold-changes and P-values that result from a DE analysis.

What you want is to be able to use the result set ID to retrieve the associated DE result, which is not currently implemented. The cleanest way to do this is to have an endpoint for result sets, regardless of the associated experiment. You definitely don't need a GEO identifier, a result set ID unambiguously refer to a single EE.

```http
GET /rest/v2/resultsets/{resultSetId}
Accept: application/json
```

This could be extended with similar search (i.e. `/rest/v2/resultsets/search`) criteria that I will implement for searching EE such as finding all result sets associated to a GEO entry, etc.

**Comment by ppavlidis** [comment]
Yes, this is the workflow we were talking about in our meeting a couple weeks back. Very roughly, we want to streamline these tasks:

1. Get the metadata for an experiment, including experimental design and pointers to DE results.
2. Get the DE results in a easy-to-use format that (and this is key) is easily "lined up" with the metadata.

It might even be reasonable under some conditions, to want to do both of those in one step, if it makes it faster.

I hope I have that right and am not making things worse on the confusion front :)

**Comment by arteymix** [comment]
I noticed that all the URLs parts are using single word. @ppavlidis is camel-casing fine for the API?

**Comment by ppavlidis** [comment]
> I noticed that all the URLs parts are using single word. @ppavlidis is camel-casing fine for the API?

I don't really mind, but we should be consistent if possible ...

**Comment by arteymix** [comment]
Indeed, but to be consistent, we need a precedent ;)
